cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/spec/PipelineSpecGenerator.java

```diff
@@ -328,41 +328,61 @@ private ConfiguredStage configureStage(ETLStage stage, ValidatedPipeline validat
+  /**
+   * Adds a Plugin usage to the Application and create a new instance.
+   *
+   * @param stageName stage name
+   * @param etlPlugin plugin
+   * @param pluginSelector plugin selector
+   * @param type type of the plugin
+   * @param pluginName plugin name
+   * @param stageConfigurer stage configurer
+   * @throws ValidationException if error while creating new plugin instance
+   * @return new instance of the plugin
+   */
+  private Object getPlugin(String stageName, ETLPlugin etlPlugin, TrackedPluginSelector pluginSelector, String type,
                            String pluginName, DefaultStageConfigurer stageConfigurer) {
     Object plugin = null;
     try {
+      // Call to usePlugin may throw IllegalArgumentException if hte plugin with the same id is already deployed.
+      // This would mean there is a bug in the app and this can not be fixed by user. That is why it is not handled as
+      // a ValidationFailure.
       plugin = pluginConfigurer.usePlugin(type, pluginName, stageName, etlPlugin.getPluginProperties(), pluginSelector);
-    } catch (IllegalArgumentException e) {
-      stageConfigurer.addFailure(e.getMessage(), String.format("Provide unique plugin id for plugin '%s'.", stageName));
-    } catch (InvalidPluginConfigException e) {
-      stageConfigurer.addFailure(e.getMessage(), "Provide valid plugin config properties.");
+    } catch (Exception e) {
+      // TODO: Catch specific exceptions when CDAP-15744 is fixed
+      stageConfigurer.addFailure(e.getMessage(), null);
     }
+
     // throw validation exception if any error occurred while creating a new instance of the plugin
     stageConfigurer.throwIfFailure();
 
     if (plugin == null) {
       String errorMessage = String.format("Plugin named '%s' of type '%s' not found.", pluginName, type);
       String correctiveAction = String.format("Make sure plugin '%s' of type '%s' is already deployed.",
                                               pluginName, type);
-      stageConfigurer.addFailure(errorMessage, correctiveAction).withPluginNotFound(stageName, pluginName, type);
+      ArtifactSelectorConfig requested = etlPlugin.getArtifactConfig();
+
+      ValidationFailure.Cause cause = stageConfigurer.addFailure(errorMessage, correctiveAction)
+        .withPluginNotFound(stageName, pluginName, type).getCauses().get(0);
+
+      cause.addAttribute("requestedArtifactName", requested.getName())
+        .addAttribute("requestedArtifactScope", requested.getScope())
+        .addAttribute("requestedArtifactVersion", requested.getVersion());
+
+      if (pluginSelector.getSuggestion() != null) {
+        // add cause attributes to suggest an artifact
+        ArtifactSelectorConfig suggestion = pluginSelector.getSuggestion();
+        cause.addAttribute("suggestedArtifactName", suggestion.getName())
+          .addAttribute("suggestedArtifactScope", suggestion.getScope())
```

Comment:
these should be in that central class with the other attributes. It's also better to add arguments to the withPluginNotFound() method to take the requested and suggested ArtifactSelectorConfigs.